### PR TITLE
Refactor preceding correction fields, using invopop/validation

### PR DESCRIPTION
**Pull request by samlown**
* Refactoring `InvoiceType` to provide a "default".
* Renaming `InvoiceTypeCorrected` to `InvoiceTypeCorrective`, including appropriate key renaming.
* Adding `Reason` field to `bill.Preceding`.
* Removing `CorrectionKey` and `CorrectionMethodKey` from `bill` package, now using `cbc.Key` directly and assuming values will be validated by regimes directly.
* Prepared example validation and correction code definitions for `es` and `co` packages.
* Defining a standard `tax.KeyDefinition` to make it easier to list regime specific definitions of keys.
* Adding `PrecedingDefinitions` to the `tax.Regime` so that it is easier to list correction key definitions for a given region.
* Replacing the ozzo-validation package with our own invopop/validation fork which corrects issues around validating using custom methods inside an array.

### CODE REVIEW

File: tax/regime.go [hunk 1]
```diff
@@ -111,6 +114,29 @@ type RateValue struct {
...
+// preceding document data.
+type PrecedingDefinitions struct {
+	// Corrections contains a list of all the keys that can be used to identify a correction.
+	Corrections []*KeyDefinition `json:"corrections,omitempty" jsonschema:"title=Corrections"`
+	// CorrectionMethods describe the methods used to correct an invoice.
+	CorrectionMethods []*KeyDefinition `json:"correction_methods,omitempty" jsonschema:"title=Correction Methods"`
+}
+
+// KeyDefinition defines properties of a key that is specific for a regime.
+type KeyDefinition struct {
```
Comment (lucanioi) [inline tax/regime.go:127]: Nice!

Comment (lucanioi) [reply tax/regime.go:127]: I'll use this to define Fattura PA codes

**Review by lucanioi — APPROVED** [review]

**Review by cavalle — APPROVED** [review]
These changes look great to me. I think they separate very nicely what is common from what is specific of each regime. Quite an improvement! 

Just left a comment to improve the list of correction methods in CO.

File: regimes/co/preceding.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		Meta: cbc.Meta{KeyDIAN: "1"},
+	},
+	{
+		Key: CorrectionMethodKeyRevoked,
+		Desc: i18n.String{
+			i18n.EN: "Previous document has been completely cancelled.",
+			i18n.ES: "Anulación de la factura anterior.",
+		},
+		Meta: cbc.Meta{KeyDIAN: "2"},
+	},
```
[the remark this answers is not in the document]
Comment (samlown) [reply regimes/co/preceding.go:32]: If they're not supported by Plemsi I think we can leave them out. I have a feeling however that we might need additional conditions such as being able to filter according to the `bill.InvoiceType`.

**Pull request merged**